### PR TITLE
Actions for Generating and Checking Jupyter Notebooks

### PR DESCRIPTION
**Pull request by lettlini**
* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [ ] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [ ] Have you added NumPy docstrings for newly added functions? 
* [ ] Have you formatted your code using black? 
* [ ] If you have introduced a new functionality, have you added adequate unit tests?
* [ ] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [x] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged? 

This pull request adds actions for
1. Generating all jupyter notebooks in `examples/` (pushing generated notebooks)
2. Checking if all jupyter notebooks in `examples/` run without any errors

### CODE REVIEW

**Comment by lettlini** [comment]
I have adjusted the trigger events according to the suggestions of @w-k-jones.
It appears that the _Jupyter Notebook CI_ action fails because there is an error in one of the documentation notebooks.
I am  not quiet sure if it is desirable to run the action on the notebooks in `doc/`. If someone could clarify this, I am happy to update this action's definition.

**Comment by freemansw1** [comment]
ha, that's amazing. Your CI discovers a bug before even being implemented. I love it. We should probably fix that. 

I think the notebooks in `doc` should also be covered by this CI.

**Comment by w-k-jones** [comment]
I think that I would be a good idea to make it run on the notebooks in `doc/` as well, to make sure that they are always up to date for the current version

**Comment by lettlini** [comment]
I agree with you @w-k-jones, I have updated the path in the corresponding `find` command accordingly.

**Comment by freemansw1** [comment]
@lettlini Are you happy for this to be reviewed and moved into an active PR? Given that your new action will fail (due to bugs in the notebooks), we should probably fix the notebook bugs before merging. That's probably a new issue and PR, and should be done before releasing v1.5.0.

**Comment by lettlini** [comment]
I think this PR is ready for being reviewed.

I agree with you @freemansw1, we should fix the errors in the notebooks before merging this PR.

**Comment by w-k-jones** [comment]
Great, awesome work @lettlini ! Do you have time to fix the notebook bugs or should someone else pick that up?

**Comment by lettlini** [comment]
I think I already found the problem. I am going to create a PR shortly

**Comment by lettlini** [comment]
It seems to have been just one notebook that was causing errors. I think I fixed the problem with #269.

**Comment by freemansw1** [comment]
Okay, with #269 merged, @lettlini if you merge the current changes, the CI should pass and we can get this in.

**Comment by lettlini** [comment]
It seems that `pytables` has been removed from the requierements file in **RC_v1.5.0**. However, the notebooks still need it to run. There are two possible solutions:
1. Add it back to the `requirements.txt` file
2. Adding an explicit install instruction in the action definition (i.e. `conda install ... pytables`)

What do you think is the best course of action?

**Comment by freemansw1** [comment]
> It seems that `pytables` has been removed from the requierements file in **RC_v1.5.0**. However, the notebooks still need it to run. There are two possible solutions:
> 
> 1. Add it back to the `requirements.txt` file
> 2. Adding an explicit install instruction in the action definition (i.e. `conda install ... pytables`)
> 
> What do you think is the best course of action?

It's funny; we had a whole conversation about this. I'm inclined for option 2, although longer-term, I'm inclined to have pandas save to non-HDF formats (well, longer term, this is irrelevant due to the xarray transition)

File: .github/workflows/generate_notebooks.yml
Comment (w-k-jones) [inline .github/workflows/generate_notebooks.yml:25]: **question:** As before, do we need a specific version?

File: .github/workflows/check_notebooks.yml
Comment (lettlini) [reply .github/workflows/check_notebooks.yml:18]: I certainly agree with you that this would be a more flexible solution. However, as far as I know, the `setup-miniconda` action can only install either the _latest_ or a _specific_ python version. In principle we could also switch to the latest python version.

Comment (freemansw1) [reply .github/workflows/check_notebooks.yml:18]: I'm inclined to use the latest python version as that's what we do in the docker CI.

**Comment by freemansw1** [comment]
> I think for now we should be keep `pytables` as a dependency simply so that the notebooks work out of the box. It will be nice when the `xarray` transition makes this unnecessary!

I'm not opposed to this, but we did explicitly remove it (#119 ) and I'm hesitant to then re-add it.

**Review by freemansw1 — APPROVED** [review]
This is excellent work, and I'm happy for this to be merged once @w-k-jones's last comments around version and `pytables` are resolved.

File: .github/workflows/check_notebooks.yml
Comment (w-k-jones) [reply .github/workflows/check_notebooks.yml:18]: Thanks for the answers! I think consistency is best, so should be change to latest to match the docker CI?

**Review by w-k-jones — COMMENTED** [review]
Great! Thanks for the fixes @lettlini , I think this is now ready to merge

**Comment by freemansw1** [comment]
@lettlini I think we're ready to merge if you are.

**Comment by w-k-jones** [comment]
Ok, potential problem here. When running the workflow after merging this the following error occurred on the `Generate-Notebooks` workflow:

```
[RC_v1.5.0 2ca348f] Generated Notebooks
 8 files changed, 90340 insertions(+), 292 deletions(-)
remote: Permission to tobac-project/tobac.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/tobac-project/tobac/': The requested URL returned error: 403
Error: Process completed with exit code 128.
```

The repository has read and write permissions enabled for workflow actions, so I'm not sure what is going on here. Perhaps the write protections are preventing the commit to the RC branch?

**Comment by lettlini** [comment]
Sorry this is causing problems! I think you are right with the assumption that this issue is caused by the branch protection rules. I simply overlooked this. 
I will look for a fix first thing tomorrow!

**Comment by freemansw1** [comment]
https://github.com/orgs/community/discussions/13836 may be a good resource here. Happy to edit the branch protections as needed for this use case.

**Comment by w-k-jones** [comment]
There is a potential workaround; we can exclude specific accounts from the protection rules. Would it be possible to make an action account for tobac and use that to submit the generated notebooks?

**Comment by lettlini** [comment]
According to the link posted by @freemansw1 that could work. I'm just not really convinced that this is a good solution.

**Comment by freemansw1** [comment]
@lettlini I'd like to understand more about your reasoning here... it seems to me like making an app under our organization with these special permissions wouldn't be too bad.

**Comment by freemansw1** [comment]
@lettlini I was thinking of making the action an app, rather than tied to a user, as is done here: https://github.com/orgs/community/discussions/13836#discussioncomment-4851440 . We should be able to make an app under the tobac-project organization and give the CI that access token without it being a new user account, I think? This is unclear to me.

**Comment by freemansw1** [comment]
I'm going to open an issue on this. I think resolving this one way or another needs to be a blocking issue before v1.5 is released. I'd like to resolve it such that we generate the notebooks through CI, but if that's not possible, we will need to manually regenerate them before release.

**Pull request merged**
